### PR TITLE
feat: add CDB64-based historical root TX index (PE-8730)

### PR DESCRIPTION
**Pull request by djwhitt**
## Summary

- Implement CDB64-based historical root TX index for efficient lookup of root transaction IDs from data item IDs
- Add comprehensive CLI tools for generating CDB64 files from CSV and SQLite sources, plus export tools
- Include full file format specification with 64-bit addressing for files up to 16 exabytes
- Ensure binary compatibility with the Rust `cdb64-rs` library for cross-platform tooling

## Key Changes

**Core Implementation:**
- `src/lib/cdb64.ts` - CDB64 reader/writer with O(1) lookups via perfect hashing
- `src/lib/cdb64-encoding.ts` - Binary encoding utilities for the index format
- `src/discovery/cdb64-root-tx-index.ts` - Runtime index with multi-file directory support and file watching

**CLI Tools:**
- `tools/generate-cdb64-root-tx-index` - Generate CDB64 from CSV files
- `tools/export-cdb64-root-tx-index` - Export CDB64 to CSV format
- `tools/export-sqlite-to-cdb64` - Export from SQLite data.db directly

**Documentation:**
- `docs/cdb64-format.md` - Complete file format specification

**Testing:**
- Unit tests for reader/writer
- Property-based tests with fast-check
- Rust interoperability tests with cdb64-rs
- CLI round-trip integration tests

## Configuration

New environment variable:
- `CDB64_ROOT_TX_INDEX_DIR` - Directory containing CDB64 index files (optional)

## Test plan

- [x] All existing tests pass
- [x] New unit tests cover reader/writer edge cases
- [x] Property-based tests verify hash consistency and round-trip correctness
- [x] Rust interoperability tests confirm cross-platform compatibility
- [x] CLI tools tested with round-trip CSV → CDB64 → CSV verification

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: tools/lib/generate-cdb64-root-tx-index-rs.ts [hunk 1]
```diff
@@ -0,0 +1,339 @@
...
+    try {
+      if (fs.existsSync(tempPath)) {
+        fs.unlinkSync(tempPath);
+      }
+    } catch {
+      // Ignore cleanup errors
+    }
+    throw error;
+  }
+}
```
[the remark this answers is not in the document]
Comment (djwhitt) [reply tools/lib/generate-cdb64-root-tx-index-rs.ts:350]: The `cdb64` package is already declared in `optionalDependencies` in package.json:

```json
"optionalDependencies": {
  "cdb64": "github:ever0de/cdb64-rs#main"
}
```

This is intentional - the Rust-backed tool is optional and will gracefully skip if the native bindings aren't available (the test file already handles this with `skip: !rustCdb64`).

File: package.json [hunk 2]
```diff
@@ -35,6 +35,7 @@
     "axios": "^1.12.2",
     "better-sqlite3": "^9.4.5",
     "change-case": "^5.4.4",
+    "chokidar": "^4.0.3",
```
[the remark this answers is not in the document]
Comment (djwhitt) [reply package.json]: Done - upgraded to `^5.0.0` in commit 57c8a7d8. All file watching tests pass.

File: src/lib/cdb64.ts [hunk 3]
```diff
@@ -0,0 +1,520 @@
...
+        }
+      }
+
+      // Move to next slot (linear probing)
+      slot = (slot + 1) % tableLength;
+    }
+
+    // Should not reach here if table is properly constructed
+    return undefined;
+  }
```
[the remark this answers is not in the document]
Comment (djwhitt) [reply src/lib/cdb64.ts:476]: Addressed in commit 30ffa021. Added `toSafeFilePosition()` helper that throws an error if file positions exceed `Number.MAX_SAFE_INTEGER` (~8TB). This provides clear feedback when CDB64 files exceed the supported size limit.

File: tools/lib/export-cdb64-root-tx-index.ts [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+  // Write header if requested
+  if (config.includeHeader) {
+    const header = 'data_item_id,root_tx_id,root_data_item_offset,root_data_offset\n';
+    if (isStdout) {
+      process.stdout.write(header);
+    } else {
+      (outputStream as fs.WriteStream).write(header);
+    }
+  }
+
```
[the remark this answers is not in the document]
Comment (djwhitt) [reply tools/lib/export-cdb64-root-tx-index.ts:156]: Addressed in commit 30ffa021. Added error event listener for output stream and check for `outputError` in the write loop to fail fast on disk errors.

File: tools/lib/generate-cdb64-root-tx-index.ts [hunk 5]
```diff
@@ -0,0 +1,305 @@
...
+            rootDataItemOffset,
+            rootDataOffset,
+          };
+          completeCount++;
+        } else {
+          // Simple format
+          value = { rootTxId };
+          simpleCount++;
+        }
+
```
[the remark this answers is not in the document]
Comment (djwhitt) [reply tools/lib/generate-cdb64-root-tx-index.ts:283]: Addressed in commit 30ffa021. Added explicit both-or-neither validation for offset columns - now throws a clear error if only one offset is provided.

File: src/lib/cdb64.rust-interop.test.ts [hunk 6]
```diff
@@ -0,0 +1,1188 @@
...
+// The cdb64-rs repo has Node bindings in the 'node' subdirectory
+let rustCdb64: typeof import('cdb64/node/index.js') | undefined;
+try {
+  rustCdb64 = await import('cdb64/node/index.js');
+} catch {
+  // Rust bindings not available - tests will be skipped
+}
+
+describe('CDB64 Rust Interoperability', { skip: !rustCdb64 }, () => {
+  let tempDir: string;
```
[the remark this answers is not in the document]
Comment (djwhitt) [reply src/lib/cdb64.rust-interop.test.ts:48]: Added clarifying comment in commit 30ffa021 explaining the import path and linking to the cdb64-rs repo. The try/catch already handles the case where the package isn't available gracefully.

**Comment by djwhitt** [comment]
Addressed validation issue in `tools/lib/export-sqlite-to-cdb64.ts` in commit b55c2f35:
- Validate that id and root_transaction_id are 32-byte Buffers before processing
- Handle partial offsets explicitly (skip instead of silently downgrading to simple format)
- Report skipped records count in summary output
- Improved error handler to handle non-Error types

**Comment by djwhitt** [comment]
Addressed double-init race condition in commit 4fe55844: Added `initializationPromise` guard to ensure concurrent calls to `getRootTx()` don't both initialize readers and watchers.

**Comment by djwhitt** [comment]
Addressed bytesRead validation in commit 1c3ed4a3: Added validation for all remaining read operations in `get()` (slot, record header, key, value) and `entries()` (key, value). All reads now throw on partial reads to detect filesystem issues or corrupted files.

File: src/lib/cdb64.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (djwhitt) [reply src/lib/cdb64.ts:476]: This has been addressed now.

**Comment by djwhitt** [comment]
Addressed watcher event handler issue in commit d4f740fb:
- Added `.catch()` to add/unlink handlers for explicit error handling at the handler level
- Fixed misleading comment about ignored callback (now says "Ignore everything except the root path and .cdb files")

File: src/discovery/cdb64-root-tx-index.ts [hunk 7]
```diff
@@ -0,0 +1,361 @@
...
+      // Key not found in any file
+      return undefined;
+    } catch (error: any) {
+      this.log.error('Error looking up root TX in CDB64', {
+        id,
+        error: error.message,
+      });
+      return undefined;
+    }
+  }
```
[the remark this answers is not in the document]
Comment (djwhitt) [reply src/discovery/cdb64-root-tx-index.ts:384]: This has been addressed.

File: src/discovery/cdb64-root-tx-index.ts [hunk 8]
```diff
@@ -0,0 +1,361 @@
...
+      await this.removeReader(filePath);
+    });
+
+    this.watcher.on('error', (error: unknown) => {
+      const message = error instanceof Error ? error.message : String(error);
+      this.log.error('CDB64 file watcher error', { error: message });
+    });
+
+    this.log.info('CDB64 file watcher started', { path: this.cdbPath });
+  }
```
[the remark this answers is not in the document]
Comment (djwhitt) [reply src/discovery/cdb64-root-tx-index.ts:142]: This has been addressed.
